php_bson.h

```diff
@@ -88,11 +88,19 @@ typedef struct {
 	{                                                                                                              \
 		{ { 0 } }, { PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL }, NULL, NULL \
 	}
+#define PHONGO_DEBUG_BSON_STATE_INITIALIZER                                                                                        \
+	{                                                                                                                              \
+		{ { 0 } }, { PHONGO_TYPEMAP_NATIVE_ARRAY, NULL, PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NATIVE_ARRAY, NULL }, NULL, NULL \
```

Comment:
Even before this PR, I think the state initializer macros were out of sync with the definitions of `php_phongo_bson_state` and `php_phongo_bson_typemap`.

If we look at https://github.com/mongodb/mongo-php-driver/commit/157481e60ff4ac06161c1f7e06fe60740616ed93#diff-079eefa18a7831e711dedc5993dba146R70 from [PHPC-922](https://jira.mongodb.org/browse/PHPC-922), the trailing `0` (for `is_visiting_array`) was replaced with `NULL`, when I think a `NULL` should have been _added_ for the new `field_path` pointer field in the `php_phongo_bson_state` struct.

Similarly, https://github.com/mongodb/mongo-php-driver/commit/bdda892b2ea3163848b4c4409d3f6968a437cad9 for [PHPC-314](https://jira.mongodb.org/browse/PHPC-314) added a new `field_paths` struct to the end of `php_phongo_bson_typemap` struct without modifying the initializer macro.

I expect this hasn't been problematic in the past, as we may not access those extra fields and/or the memory happens to be cleared; however, I think this would be a good opportunity to clear this up.

Alternatively, you may want to replace all of this with two macros. One that calls `memset()` on the `php_phongo_bson_state` struct to clear it out (as in [this thread](https://stackoverflow.com/a/11152208/162228)) and another that assigns specific fields for the debug type map. Since `PHONGO_TYPEMAP_NONE` is the first element in its enum, we can rely on it being equal to zero (see: [this thread](https://stackoverflow.com/q/6434105/162228)).

This approach would also allow us to remove the conditional definitions for PHP 5 and 7, since all we really care about doing is assigning two type map fields. Consider:

```c
#define INIT_PHONGO_BSON_STATE(s) \
    do { \
        memset(&(s), 0, sizeof(php_phongo_bson_state));
    } while (0)
#define INIT_PHONGO_DEBUG_BSON_STATE(s) \
    do { \
        memset(&(s), 0, sizeof(php_phongo_bson_state)); \
        s.map.root_type     = PHONGO_TYPEMAP_NATIVE_ARRAY; \
        s.map.document_type = PHONGO_TYPEMAP_NATIVE_ARRAY; \
    } while (0)
```

Thoughts?

For extended reading, I can point out that these macros have historically been a pain in the tuchus. See:

 * https://github.com/mongodb/mongo-php-driver/commit/321c577f982e79459fdef07de0a047cfa514e741
 * https://github.com/mongodb/mongo-php-driver/commit/a869c9746a0e25306f81ca0dddffaa12bef63a7f
 * https://github.com/mongodb/mongo-php-driver/pull/591

Reply:
That looks a lot cleaner, as it doesn't require us to continuously modify the initialiser when adding to the struct. Unfortunately, the syntax isn't as easy as previously where we could initialise the struct on declaration, but the additional call isn't that much of a problem.